colcon_sanitizer_reports/sanitizer_log_parser.py

```diff
@@ -0,0 +1,195 @@
+from collections import defaultdict
+import csv
+from io import StringIO
+import re
+from typing import Dict, List, Optional, Tuple
+
+
+class _KeysFactory:
+    """Finds key stack trace lines in sanitizer error subsections."""
+
+    @staticmethod
```

Comment:
Instead of making this static, what about making this a top-level function?

Reply:
Mostly wanted to collect these static functions in this class to make it more obvious they're closely coupled. Wasn't sure it would be clear otherwise.

Reply:
If you have a group of closely related functions which are not sharing any state, it seems to me the definition of a module (i.e. a different file containing only those functions).

Reply:
Otherwise a prefix could work too but it seems to me modules are better because if there is no risk of confusion, the prefix can be removed:

```py
def pkg_foo_xyz():
  pass

# vs.
from pkg import foo

foo.xyz()
```

Reply:
While cleaning this up and fixing other documentation, realized that I can make it more obvious what is happening here and set us up for a better time later on. We're really trying to find specific stack traces, and then a single key line from within each stack trace. This code doesn't make that very clear.

I'm going to take your suggestion of moving this into another module, but call it something like sanitizer_stack_trace.py. End result will be the same csv, but I think it'll be more clear and will set us up better for reporting information about stack traces if we eventually pursue that.

Additionally, if I'm making a module for sanitizer_stack_trace.py, it follows that "_Section" and "_SubSection" should get the same treatment.

Reply:
SGTM, I'm not sure but I think you can call something `_foo.py` to make very clear it's a private module. Not sure if the linter is going to like that or not.